OSDOCS-4347: Account Information: installing-aws-localzone created from cloudformation (#54535)
openshift/openshift-docs: installing/installing_aws/installing-aws-localzone.adoc

:_content-type: ASSEMBLY
[id="installing-aws-localzone"]
= Installing a cluster using AWS Local Zones
include::_attributes/common-attributes.adoc[]
:context: installing-aws-localzone

toc::[]

In {product-title} version {product-version}, you can install a cluster on Amazon Web Services (AWS) into an existing VPC, extending workers to the edge of the Cloud Infrastructure using AWS Local Zones.

AWS Local Zones are a type of infrastructure that place Cloud Resources close to the metropolitan regions. For more information, see the link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-local-zones[AWS Local Zones Documentation].

{product-title} can be installed in existing VPCs with Local Zone subnets. The Local Zone subnets can be used to extend the regular workers' nodes to the edge networks. The edge worker nodes are dedicated to running user workloads.

One way to create the VPC and subnets is to use the provided CloudFormation templates. You can modify the templates to customize your infrastructure or use the information that they contain to create AWS objects according to your company's policies.

[IMPORTANT]
====

> The steps for performing an installer-provisioned infrastructure installation are provided as an example only. Installing a cluster with VPC you provide requires knowledge of the cloud provider and the installation process of {product-title}. The CloudFormation templates are provided to assist in completing these steps or to help model your own. You are also free to create the required resources through other methods; the templates are just an example.

Reviewer: ```suggestion
The steps for performing an installer-provisioned infrastructure installation are provided as an example only. Installing a cluster with a VPC you provide requires knowledge of the cloud provider and the installation process of {product-title}. The CloudFormation templates are provided to assist in completing these steps or to help model your own. You are also free to create the required resources through other methods; the templates are just an example.
```

> The steps for performing an installer-provisioned infrastructure installation are provided as an example only. Installing a cluster with VPC you provide requires knowledge of the cloud provider and the installation process of {product-title}. The CloudFormation templates are provided to assist in completing these steps or to help model your own. You are also free to create the required resources through other methods; the templates are just an example.

Reviewer 2: Yes. This is more IPI than UPI. However, it does use some CF templates to make the VPC and subnets. Unlike the UPI methods, the installation program still stands up all the machines. The VPC and subnet data is added to a custom Kubernetes manifest for the workers to get it done. I think the note still holds.

* You xref:../../installing/installing_aws/installing-aws-account.adoc#installing-aws-account[configured an AWS account] to host the cluster.
+
[IMPORTANT]
====
If you have an AWS profile stored on your computer, it must not use a temporary session token that you generated while using a multi-factor authentication device. The cluster continues to use your current AWS credentials to create AWS resources for the entire life of the cluster, so you must use key-based, long-lived credentials. To generate appropriate keys, see link:https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html[Managing Access Keys for IAM Users] in the AWS documentation. You can supply the keys when you run the installation program.
====
* You noted the region and supported link:https://aws.amazon.com/about-aws/global-infrastructure/localzones/locations[AWS Local Zones locations] to create the network resources in.
* You read the link:https://aws.amazon.com/about-aws/global-infrastructure/localzones/features/[Features] for each AWS Local Zones location

> * You downloaded the AWS CLI and installed it on your computer. See link:https://docs.aws.amazon.com/cli/latest/userguide/install-bundle.html[Install the AWS CLI Using the Bundled Installer (Linux, macOS, or UNIX)] in the AWS documentation.

Reviewer 2: I'm going to use https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html. This link is used in five other places in the repo.

:_content-type: ASSEMBLY
[id="installing-aws-localzone"]

> = AWS extending workers to the Local Zones

Author: This currently reads:

= Installing a cluster on user-provisioned infrastructure in AWS using Local Zones

include::_attributes/common-attributes.adoc[]
:context: installing-aws-localzone

toc::[]

In {product-title} version {product-version}, you can install a cluster on Amazon Web Services (AWS) that uses both infrastructure that you provide and Local Zones.

> One way to create this infrastructure is to use the provided CloudFormation templates. You can modify the templates to customize your infrastructure or use the information that they contain to create AWS objects according to your company's policies.

Reviewer 3: 👍

[IMPORTANT]
====

> The steps for performing a user-provisioned infrastructure installation are provided as an example only. Installing a cluster with infrastructure you provide requires knowledge of the cloud provider and the installation process of {product-title}. Several CloudFormation templates are provided to assist in completing these steps or to help model your own. You are also free to create the required resources through other methods; the templates are just an example.

Author: Done

====
* You downloaded the AWS CLI and installed it on your computer. See link:https://docs.aws.amazon.com/cli/latest/userguide/install-bundle.html[Install the AWS CLI Using the Bundled Installer (Linux, macOS, or UNIX)] in the AWS documentation.
* If you use a firewall, you xref:../../installing/install_config/configuring-firewall.adoc#configuring-firewall[configured it to allow the sites] that your cluster requires access to.
+
[NOTE]
====
Be sure to also review this site list if you are configuring a proxy.
====
* If the cloud identity and access management (IAM) APIs are not accessible in your environment, or if you do not want to store an administrator-level credential secret in the `kube-system` namespace, you can xref:../../installing/installing_aws/manually-creating-iam.adoc#manually-creating-iam-aws[manually create and maintain IAM credentials].

include::modules/cluster-entitlements.adoc[leveloffset=+1]

[id="installation-requirements-user-infra_{context}"]
== Requirements for a cluster with installer-provisioned infrastructure

> For a cluster that contains installer-provisioned infrastructure, the Machine API will be responsible for creating the worker nodes declared by the MachineSets available on `<install-dir>/openshift/99_openshift-cluster-api_worker-machineset-*.yaml`

Author: Removed.

This section describes the requirements for deploying {product-title} on an existing VPC with Local Zone subnets.

include::modules/installation-machine-requirements.adoc[leveloffset=+2]
include::modules/installation-minimum-resource-requirements.adoc[leveloffset=+2]

> include::modules/installation-aws-tested-machine-types.adoc[leveloffset=+2]

Reviewer 3: The types supported by Local Zone are not the same to this list, only few instance types are available [1] in Local Zone, and the types supported by each Local Zone are also different. At present, only `m5.xlarge` are tested by QE. 
I would suggest remove this section or adding a new list for Local Zone separately. 

@patrickdillon @mtulio WDYT?

[1] https://aws.amazon.com/about-aws/global-infrastructure/localzones/features/?pg=localzones

Reviewer 4: > The types supported by Local Zone are not the same to this list, only few instance types are available [1] in Local Zone, and the types supported by each Local Zone are also different. At present, only `m5.xlarge` are tested by QE. I would suggest remove this section or adding a new list for Local Zone separately.
> 
> @patrickdillon @mtulio WDYT?
> 
> [1] https://aws.amazon.com/about-aws/global-infrastructure/localzones/features/?pg=localzones

@yunjiang29 I agree to add a separate list. I would suggest validating at least three instance types. Some time ago I exported the matrix of instances supported. As the instances can vary between zones, I'd suggest to validated those instances:
- `m5.xlarge`
- `m5.2xlarge`
- `c5d.2xlarge`
- `t3.xlarge`

https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42

Reviewer 3: @mtulio To be clear, `supported` doesn't equal to `tested`, for example, there are tons of instance types in various AWS regions, but we just tested some of them which are listed in the documentation. (that's why we call it `Tested instance types for AWS`, but user can use any instance types which meet [minimum resource requirements](https://docs.openshift.com/container-platform/4.11/installing/installing_aws/installing-aws-customizations.html#installation-minimum-resource-requirements_installing-aws-customizations) )

I will do validation on above instance types, and create PR on installer side to add a new list next Monday, so that this PR can include it. Following the current rule, the list would be:

```
m5.*
c5d.*
t3.*
```

cc @GroceryBoyJr @kalexand-rh

Reviewer 2: For now, I'm making a manual edit to show `m5.xlarge`, since that is true right now, and we can replace the list as soon as that PR clears.

> include::modules/installation-aws-arm-tested-machine-types.adoc[leveloffset=+2]

Reviewer 3: ARM instance types are not supported in Local Zone, please remove this section.

Author: Removed:
include::modules/installation-aws-arm-tested-machine-types.adoc[leveloffset=+2]

[role="_additional-resources"]
.Additional resources

* You can view details about the CloudFormation stacks that you create by navigating to the link:https://console.aws.amazon.com/cloudformation/[AWS CloudFormation console].

include::modules/installation-creating-aws-subnet-localzone.adoc[leveloffset=+1]

include::modules/installation-cloudformation-subnet-localzone.adoc[leveloffset=+2]

include::modules/installation-obtaining-installer.adoc[leveloffset=+1]

include::modules/ssh-agent-using.adoc[leveloffset=+1]

include::modules/installation-user-infra-generate.adoc[leveloffset=+1]

include::modules/installation-generate-aws-user-infra-install-config.adoc[leveloffset=+2]

[role="_additional-resources"]
.Additional resources

* See link:https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html[Configuration and credential file settings] in the AWS documentation for more information about AWS profile and credential configuration.

//include::modules/installation-configure-proxy.adoc[leveloffset=+2]
//Put this back if QE validates it.

include::modules/installation-user-infra-generate-k8s-manifest-ignition.adoc[leveloffset=+2]

> include::modules/installation-extracting-infraid.adoc[leveloffset=+1]

Reviewer 4: I can't see reasons to keep this information, unless we are referencing it somewhere I can't see in my review. But it isn't a requirement for Local Zones
```suggestion
```

Author: Done, thank you!

> 

Reviewer 4: After set all the manifests, we can create the cluster. I am not sure if there is a module for that, but it's a single command:
```
./openshift-install create cluster --log-level=debug
```

https://github.com/mtulio/mtulio.labs/blob/article-ocp-aws-lz/docs/articles/ocp-aws-local-zones-day-0.md#install-the-cluster-

One way to create this infrastructure is to use the provided CloudFormation templates. You can modify the templates to customize your infrastructure or use the information that they contain to create AWS objects according to your company's policies.

[IMPORTANT]
====
The steps for performing a user-provisioned infrastructure installation are provided as an example only. Installing a cluster with infrastructure you provide requires knowledge of the cloud provider and the installation process of {product-title}. Several CloudFormation templates are provided to assist in completing these steps or to help model your own. You are also free to create the required resources through other methods; the templates are just an example.
====

== Prerequisites

* You reviewed details about the xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update] processes.
* You read the documentation on xref:../../installing/installing-preparing.adoc#installing-preparing[selecting a cluster installation method and preparing it for users].

> * You xref:../../installing/installing_aws/installing-aws-account.adoc#installing-aws-account[configured an AWS account] to host the cluster and configured local zones for your account.

Reviewer 4: I believe we can add a directly link to the Opt In section, if you decide to keep the opt in on the "Aws account" page:
```suggestion
* You xref:../../installing/installing_aws/installing-aws-account.adoc#installing-aws-account[configured an AWS account] to host the cluster
* You xref:../../installing/installing_aws/installing-aws-account.adoc#installation-aws-add-local-zone-locations_installing-aws-account[opted in the AWS Local Zone group] which the subnets will be created.
```

Author: Done. The exact addition I provided is: 

* You xref:../../installing/installing_aws/installing-aws-account.adoc#installing-aws-account[configured an AWS account] to host the cluster. 
* You xref:../../installing/installing_aws/installing-aws-account.adoc#installion-aws-add-local-zone-locations_installing-aws-account[opted in to the AWS Local Zone group] in which the subnets will be created.

include::modules/installation-minimum-resource-requirements.adoc[leveloffset=+2]
include::modules/installation-aws-tested-machine-types.adoc[leveloffset=+2]
include::modules/installation-aws-arm-tested-machine-types.adoc[leveloffset=+2]

include::modules/installation-aws-marketplace-subscribe.adoc[leveloffset=+1]

//These next four modules seem like they're placed awkwardly, but you need the subnets to put in the install-config.yaml file.
include::modules/installation-creating-aws-vpc-localzone.adoc[leveloffset=+1]

include::modules/installation-cloudformation-vpc-localzone.adoc[leveloffset=+2]

[role="_additional-resources"]
.Additional resources

* You can view details about the CloudFormation stacks that you create by navigating to the link:https://console.aws.amazon.com/cloudformation/[AWS CloudFormation console].

include::modules/installation-creating-aws-subnet-localzone.adoc[leveloffset=+1]

include::modules/installation-cloudformation-subnet-localzone.adoc[leveloffset=+2]

include::modules/installation-obtaining-installer.adoc[leveloffset=+1]

include::modules/ssh-agent-using.adoc[leveloffset=+1]

include::modules/installation-user-infra-generate.adoc[leveloffset=+1]

> include::modules/installation-disk-partitioning-upi-templates.adoc[leveloffset=+2]

Reviewer 4: I think this optional also can be removed.
```suggestion
```

Author: Deleted, thank you!

====
* You downloaded the AWS CLI and installed it on your computer. See link:https://docs.aws.amazon.com/cli/latest/userguide/install-bundle.html[Install the AWS CLI Using the Bundled Installer (Linux, macOS, or UNIX)] in the AWS documentation.
* If you use a firewall, you xref:../../installing/install_config/configuring-firewall.adoc#configuring-firewall[configured it to allow the sites] that your cluster requires access to.
+
[NOTE]
====
Be sure to also review this site list if you are configuring a proxy.
====
* If the cloud identity and access management (IAM) APIs are not accessible in your environment, or if you do not want to store an administrator-level credential secret in the `kube-system` namespace, you can xref:../../installing/installing_aws/manually-creating-iam.adoc#manually-creating-iam-aws[manually create and maintain IAM credentials].

include::modules/cluster-entitlements.adoc[leveloffset=+1]

[id="installation-requirements-user-infra_{context}"]

> == Requirements for a cluster with user-provisioned infrastructure

Author: I've changed this to "Requirements for a cluster with installer-provisioned infrastructure"

For a cluster that contains user-provisioned infrastructure, you must deploy all

> of the required machines.

Author: I've edited this sentence to become: "For a cluster that contains installer-provisioned infrastructure, the Machine API will be responsible for creating the worker nodes declared by the MachineSets available on `<install-dir>/openshift/99_openshift-cluster-api_worker-machineset-*.yaml` " (the single quote is asciidoc markup)

> This section describes the requirements for deploying {product-title} on user-provisioned infrastructure.

Author: @mtulio I changed this to "This section describes the requirements for deploying {product-title} on an existing AWS VPC with Local Zone subnets." Is this OK?

> This section describes the requirements for deploying {product-title} on user-provisioned infrastructure.

Reviewer 4: @GroceryBoyJr lgtm

include::modules/installation-machine-requirements.adoc[leveloffset=+2]
include::modules/installation-minimum-resource-requirements.adoc[leveloffset=+2]
include::modules/installation-aws-tested-machine-types.adoc[leveloffset=+2]
include::modules/installation-aws-arm-tested-machine-types.adoc[leveloffset=+2]

> include::modules/csr-management.adoc[leveloffset=+2]

Author: Done! Thank you!

> include::modules/installation-supported-aws-machine-types.adoc[leveloffset=+2]

Author: I've deleted line 53, "include::modules/installation-supported-aws-machine-types.adoc"

include::modules/installation-aws-user-infra-requirements.adoc[leveloffset=+1]

> include::modules/installation-aws-permissions.adoc[leveloffset=+2]

Author: I have deleted lines 54-57.

include::modules/installation-aws-marketplace-subscribe.adoc[leveloffset=+1]

> //These next four modules seem like they're placed awkwardly, but you need the subnets to put in the install-config.yaml file.

Reviewer 2: Is this placement correct?

Reviewer 2: As long as it's in the right place in the process, I think it would be better to put them like this.

* See xref:../../installing/installing_aws/installing-aws-user-infra.adoc#installation-aws-user-infra-rhcos-ami_installing-aws-user-infra[{op-system} AMIs for the AWS infrastructure] for details about the {op-system-first} AMIs for the AWS zones.

include::modules/installation-creating-aws-control-plane.adoc[leveloffset=+1]

include::modules/installation-cloudformation-control-plane.adoc[leveloffset=+2]

[role="_additional-resources"]
.Additional resources

* You can view details about the CloudFormation stacks that you create by navigating to the link:https://console.aws.amazon.com/cloudformation/[AWS CloudFormation console].

include::modules/installation-creating-aws-worker.adoc[leveloffset=+1]

////
[id="installing-workers-aws-user-infra"]
== Creating worker nodes

You can either manually create worker nodes or use a MachineSet to create worker
nodes after the cluster deploys. If you use a MachineSet to create and maintain
the workers, you can allow the cluster to manage them. This allows you to easily
scale, manage, and upgrade your workers.
////

include::modules/installation-cloudformation-worker.adoc[leveloffset=+2]

> //Do we need to remove the instructions to create separate worker nodes? It looks like the installation program might make the workers because we're creating a manifest for them.

Reviewer 2: I am very curious about this.

Reviewer 4: Yes, we need to remove it. I left a comment with suggested changes to remove those items.

Author: I have removed this include: plus the comment.

You can configure registry storage for user-provisioned infrastructure in AWS to deploy {product-title} to hidden regions. See xref:../../registry/configuring_registry_storage/configuring-registry-storage-aws-user-infrastructure.adoc#configuring-registry-storage-aws-user-infrastructure[Configuring the registry for AWS user-provisioned infrastructure] for more information.

include::modules/registry-configuring-storage-aws-user-infra.adoc[leveloffset=+3]

include::modules/installation-registry-storage-non-production.adoc[leveloffset=+3]

include::modules/installation-aws-user-infra-delete-bootstrap.adoc[leveloffset=+1]

include::modules/installation-create-ingress-dns-records.adoc[leveloffset=+1]

include::modules/installation-aws-user-infra-installation.adoc[leveloffset=+1]

include::modules/logging-in-by-using-the-web-console.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* See xref:../../web_console/web-console.adoc#web-console[Accessing the web console] for more details about accessing and understanding the {product-title} web console.

include::modules/cluster-telemetry.adoc[leveloffset=+1]

[role="_additional-resources"]

> .Additional resources

Reviewer 4: I think we can add a few resources here, something like:
- ["Opt in to Local Zones"](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#opt-in-local-zone)
- Local Zones available locations https://aws.amazon.com/about-aws/global-infrastructure/localzones/locations
- Local Zones features https://aws.amazon.com/about-aws/global-infrastructure/localzones/features

Reviewer 2: I'm moving these links down to the `== Additional resources` section down a few lines because they apply to the whole document.

* See xref:../../support/remote_health_monitoring/about-remote-health-monitoring.adoc#about-remote-health-monitoring[About remote health monitoring] for more information about the Telemetry service.

[role="_additional-resources"]
[id="installing-localzone-additional-resources"]
== Additional resources

* See link:https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stacks.html[Working with stacks] in the AWS documentation for more information about AWS CloudFormation stacks.

[id="installing-aws-localzone-next-steps"]

> == Next steps

Reviewer 4: @kalexand-rh @GroceryBoyJr 
I am thinking if we need to provide a section, or reference, of deploying a simple app in Local Zone worker nodes, which require the user to define the tolerations on the deployment spec. I will check if we have something in our documentation to make the reference. Let me know if you consider it a requirement for now or we can patch it later.

Reviewer 2: @mtulio, I think that we can patch that in after GA. Let me know if you have a reference doc, and I'll make an issue to track the work.

Reviewer 4: x-ref: https://issues.redhat.com/browse/OSDOCS-4812
